File: packages/workbench-model/src/contribution-calendar.ts
export type ContributionMetric = 'activity' | 'tokens'

export type ContributionPoint = {
  day: string
  value: number
}

export type ContributionCalendarDay = ContributionPoint & {
  intensity: 0 | 1 | 2 | 3 | 4
  isFuture: boolean
  date: Date
}

export type ContributionCalendarWeek = {
  startDay: string
  days: ContributionCalendarDay[]
}

export type ContributionMonthLabel = {
  date: Date
  weekIndex: number
}

export type ContributionCalendar = {
  weeks: ContributionCalendarWeek[]
  monthLabels: ContributionMonthLabel[]
  maxValue: number
}

export type ContributionTotals = {
  today: number
  currentStreak: number
  longestStreak: number
  visibleTotal: number
}

const CALENDAR_WEEK_COUNT = 53
const DAYS_PER_WEEK = 7

export function buildContributionCalendar(
  points: readonly ContributionPoint[],
  anchorDate = new Date()
): ContributionCalendar {
  const valuesByDay = mergeContributionPoints(points)
  const anchor = startOfLocalDay(anchorDate)
  const currentWeekStart = addLocalDays(anchor, -anchor.getDay())
  const gridStart = addLocalDays(currentWeekStart, -(CALENDAR_WEEK_COUNT - 1) * DAYS_PER_WEEK)
  const firstVisibleDay = localCalendarDayKey(gridStart)
  const lastVisibleDay = localCalendarDayKey(anchor)
  let maxValue = 0
  for (const [day, value] of valuesByDay) {
    if (day >= firstVisibleDay && day <= lastVisibleDay) {
      maxValue = Math.max(maxValue, value)
    }
  }
  const weeks: ContributionCalendarWeek[] = []

  for (let weekIndex = 0; weekIndex < CALENDAR_WEEK_COUNT; weekIndex++) {
    const days: ContributionCalendarDay[] = []
    for (let weekday = 0; weekday < DAYS_PER_WEEK; weekday++) {
      const date = addLocalDays(gridStart, weekIndex * DAYS_PER_WEEK + weekday)
      const day = localCalendarDayKey(date)
      const value = Math.max(0, valuesByDay.get(day) ?? 0)
      days.push({
        day,
        value,
        intensity: contributionIntensity(value, maxValue),
        isFuture: date.getTime() > anchor.getTime(),
        date
      })
    }
    weeks.push({ startDay: days[0]?.day ?? localCalendarDayKey(gridStart), days })
  }

  return {
    weeks,
    monthLabels: buildMonthLabels(weeks),
    maxValue
  }
}

export function getContributionTotals(
  points: readonly ContributionPoint[],
  anchorDate = new Date()
): ContributionTotals {
  const valuesByDay = mergeContributionPoints(points)
  const anchor = startOfLocalDay(anchorDate)
  const today = valuesByDay.get(localCalendarDayKey(anchor)) ?? 0
  let currentStreak = 0
  let longestStreak = 0
  let runningStreak = 0
  let visibleTotal = 0

  for (let offset = 365; offset >= 0; offset--) {
    const value = valuesByDay.get(localCalendarDayKey(addLocalDays(anchor, -offset))) ?? 0
    visibleTotal += value
    if (value > 0) {
      runningStreak++
      longestStreak = Math.max(longestStreak, runningStreak)
    } else {
      runningStreak = 0
    }
  }

  const streakAnchor = today > 0 ? anchor : addLocalDays(anchor, -1)
  for (let offset = 0; offset <= 365; offset++) {
    const value = valuesByDay.get(localCalendarDayKey(addLocalDays(streakAnchor, -offset))) ?? 0
    if (value <= 0) {
      break
    }
    currentStreak++
  }

  return { today, currentStreak, longestStreak, visibleTotal }
}

function buildMonthLabels(weeks: ContributionCalendarWeek[]): ContributionMonthLabel[] {
  const labels: ContributionMonthLabel[] = []
  for (let weekIndex = 0; weekIndex < weeks.length; weekIndex++) {
    const firstOfMonth = weeks[weekIndex]?.days.find(
      (day) => day.date.getDate() === 1 && !day.isFuture
    )
    if (firstOfMonth) {
      labels.push({ date: firstOfMonth.date, weekIndex })
    }
  }
  if (labels[0]?.weekIndex !== 0) {
    const firstDay = weeks[0]?.days.find((day) => !day.isFuture)
    if (firstDay) {
      labels.unshift({ date: firstDay.date, weekIndex: 0 })
    }
  }
  return labels
}

function mergeContributionPoints(points: readonly ContributionPoint[]): Map<string, number> {
  const valuesByDay = new Map<string, number>()
  for (const point of points) {
    if (!Number.isFinite(point.value) || point.value <= 0) {
      continue
    }
    valuesByDay.set(point.day, (valuesByDay.get(point.day) ?? 0) + point.value)
  }
  return valuesByDay
}

function contributionIntensity(value: number, maxValue: number): 0 | 1 | 2 | 3 | 4 {
  if (value <= 0 || maxValue <= 0) {
    return 0
  }
  const scaled = Math.ceil((Math.log1p(value) / Math.log1p(maxValue)) * 4)
  if (scaled <= 1) {
    return 1
  }
  if (scaled === 2) {
    return 2
  }
  if (scaled === 3) {
    return 3
  }
  return 4
}

function startOfLocalDay(value: Date): Date {
  const date = new Date(value)
  date.setHours(0, 0, 0, 0)
  return date
}

function addLocalDays(value: Date, amount: number): Date {
  const date = new Date(value)
  date.setDate(date.getDate() + amount)
  return date
}

export function localCalendarDayKey(value: Date | number): string {
  const date = typeof value === 'number' ? new Date(value) : value
  const year = date.getFullYear()
  const month = String(date.getMonth() + 1).padStart(2, '0')
  const day = String(date.getDate()).padStart(2, '0')
  return `${year}-${month}-${day}`
}

File: apps/mobile-ios/YiruMobile/Features/Activity/contribution-calendar.swift
import Foundation
import SwiftUI

nonisolated struct ActivityContributionDay: Hashable, Identifiable, Sendable {
    let day: String
    let value: Double
    let intensity: Int
    let isFuture: Bool
    let date: Date
    var id: String { day }
}

nonisolated struct ActivityContributionWeek: Hashable, Identifiable, Sendable {
    let startDay: String
    let days: [ActivityContributionDay]
    var id: String { startDay }
}

nonisolated struct ActivityContributionTotals: Sendable {
    let today: Double
    let currentStreak: Int
    let longestStreak: Int
    let visibleTotal: Double
}

nonisolated func activityContributionWeeks(
    _ points: [ActivityDailyPoint],
    metric: ActivityMetric,
    now: Date = Date()
) -> [ActivityContributionWeek] {
    let calendar = Calendar.current
    let anchor = calendar.startOfDay(for: now)
    let weekday = calendar.component(.weekday, from: anchor)
    let currentWeekStart =
        calendar.date(byAdding: .day, value: -(weekday - 1), to: anchor) ?? anchor
    let gridStart =
        calendar.date(byAdding: .day, value: -(52 * 7), to: currentWeekStart)
        ?? currentWeekStart
    let values = contributionValues(points, metric: metric)
    let firstVisible = activityDayKey(gridStart)
    let lastVisible = activityDayKey(anchor)
    let maximum = values.reduce(0.0) { result, entry in
        entry.key >= firstVisible && entry.key <= lastVisible ? max(result, entry.value) : result
    }
    return (0..<53).map { weekIndex in
        let days = (0..<7).map { weekdayIndex in
            let offset = weekIndex * 7 + weekdayIndex
            let date = calendar.date(byAdding: .day, value: offset, to: gridStart) ?? gridStart
            let day = activityDayKey(date)
            let value = max(0, values[day] ?? 0)
            return ActivityContributionDay(
                day: day,
                value: value,
                intensity: contributionIntensity(value, maximum: maximum),
                isFuture: date > anchor,
                date: date
            )
        }
        return ActivityContributionWeek(startDay: days[0].day, days: days)
    }
}

nonisolated func activityContributionTotals(
    _ points: [ActivityDailyPoint],
    metric: ActivityMetric = .activity,
    now: Date = Date()
) -> ActivityContributionTotals {
    let calendar = Calendar.current
    let anchor = calendar.startOfDay(for: now)
    let values = contributionValues(points, metric: metric)
    let today = values[activityDayKey(anchor)] ?? 0
    var longest = 0
    var running = 0
    var visibleTotal = 0.0
    for offset in stride(from: 365, through: 0, by: -1) {
        let date = calendar.date(byAdding: .day, value: -offset, to: anchor) ?? anchor
        let value = values[activityDayKey(date)] ?? 0
        visibleTotal += value
        if value > 0 {
            running += 1
            longest = max(longest, running)
        } else {
            running = 0
        }
    }
    let streakAnchor =
        today > 0
        ? anchor : calendar.date(byAdding: .day, value: -1, to: anchor) ?? anchor
    var current = 0
    for offset in 0...365 {
        let date = calendar.date(byAdding: .day, value: -offset, to: streakAnchor) ?? streakAnchor
        guard (values[activityDayKey(date)] ?? 0) > 0 else { break }
        current += 1
    }
    return ActivityContributionTotals(
        today: today,
        currentStreak: current,
        longestStreak: longest,
        visibleTotal: visibleTotal
    )
}

nonisolated private func contributionValues(
    _ points: [ActivityDailyPoint],
    metric: ActivityMetric
) -> [String: Double] {
    var values: [String: Double] = [:]
    for point in points {
        let value = activityMetricValue(metric, daily: point)
        guard value.isFinite, value > 0 else { continue }
        values[point.day, default: 0] += value
    }
    return values
}

nonisolated private func contributionIntensity(_ value: Double, maximum: Double) -> Int {
    guard value > 0, maximum > 0 else { return 0 }
    return min(4, max(1, Int(ceil((log1p(value) / log1p(maximum)) * 4))))
}

nonisolated func activityDayKey(_ date: Date) -> String {
    let components = Calendar.current.dateComponents([.year, .month, .day], from: date)
    return String(
        format: "%04d-%02d-%02d",
        components.year ?? 0,
        components.month ?? 0,
        components.day ?? 0
    )
}

struct ActivityContributionCard: View {
    let points: [ActivityDailyPoint]
    @Binding var metric: ActivityMetric
    @State private var selected: ActivityContributionDay?

    var body: some View {
        ContentSurface {
            VStack(alignment: .leading, spacing: 12) {
                HStack(alignment: .top, spacing: 12) {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("Contribution history")
                            .font(.system(size: 14, weight: .semibold))
                        Text(description)
                            .font(.system(size: 12))
                            .foregroundStyle(Theme.Colors.mutedForeground)
                    }
                    Spacer(minLength: 0)
                    // Why: a narrow two-way Activity/Tokens control in the card header. The
                    // third "API value" state is reachable only by tapping the trend/rhythm
                    // charts below, never as a visible segment, so it stays out of this picker.
                    Picker("Contribution metric", selection: cardMetric) {
                        Text(ActivityMetric.activity.title).tag(ActivityMetric.activity)
                        Text(ActivityMetric.tokens.title).tag(ActivityMetric.tokens)
                    }
                    .pickerStyle(.segmented)
                    .frame(width: 144)
                    .accessibilityLabel("Contribution metric")
                }
                ScrollViewReader { proxy in
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(alignment: .top, spacing: 4) {
                            weekdayLabels
                            ForEach(weeks) { week in
                                VStack(spacing: 4) {
                                    ForEach(week.days) { day in
                                        contributionCell(day)
                                    }
                                }
                                .id(week.id)
                            }
                        }
                        .padding(.top, 16)
                    }
                    .onAppear {
                        if let last = weeks.last { proxy.scrollTo(last.id, anchor: .trailing) }
                    }
                }
                HStack(spacing: 8) {
                    Text(footer)
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Colors.mutedForeground)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    Text("Less")
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Colors.mutedForeground)
                    ForEach(0..<5, id: \.self) { intensity in
                        Rectangle()
                            .fill(color(intensity))
                            .frame(width: 8, height: 8)
                            .overlay { Rectangle().stroke(.separator.opacity(0.6), lineWidth: 0.5) }
                    }
                    Text("More")
                        .font(.system(size: 11))
                        .foregroundStyle(Theme.Colors.mutedForeground)
                }
            }
            .padding(16)
        }
        .onChange(of: metric) { _, _ in selected = nil }
    }

    private var weeks: [ActivityContributionWeek] {
        activityContributionWeeks(points, metric: metric)
    }

    // Why: the header toggle shows Tokens selected while the hidden `.value` state is
    // active, and always writes back an explicit `.activity`/`.tokens` choice so the
    // two-segment control never has to represent a third state.
    private var cardMetric: Binding<ActivityMetric> {
        Binding(
            get: { metric == .activity ? .activity : .tokens },
            set: { metric = $0 }
        )
    }

    private var weekdayLabels: some View {
        VStack(spacing: 4) {
            ForEach(0..<7, id: \.self) { weekday in
                Text(weekday % 2 == 1 ? Calendar.current.veryShortWeekdaySymbols[weekday] : "")
                    .font(.system(size: 9))
                    .foregroundStyle(Theme.Colors.mutedForeground)
                    .frame(width: 12, height: 10)
            }
        }
        .padding(.trailing, 1)
    }

    // Why: day-cell selection only means something while browsing Activity. Once
    // Tokens/API value is active, tapping any cell cycles the shared metric instead — the
    // same gesture the trend/rhythm/provider charts use — and `onChange(of: metric)` above
    // clears the selection so a stale day cannot outlive the metric it was read under.
    private func contributionCell(_ day: ActivityContributionDay) -> some View {
        Button {
            guard !day.isFuture else { return }
            if metric == .activity {
                selected = day
            } else {
                metric = nextTokenValueMetric(metric)
            }
        } label: {
            Rectangle()
                .fill(day.isFuture ? .clear : color(day.intensity))
                .frame(width: 10, height: 10)
                .overlay {
                    Rectangle().stroke(
                        selected?.day == day.day
                            ? Theme.Colors.foreground
                            : Theme.Colors.selection.opacity(day.isFuture ? 0 : 0.6),
                        lineWidth: selected?.day == day.day ? 1 : 0.5
                    )
                }
        }
        .buttonStyle(.plain)
        .disabled(day.isFuture)
        .accessibilityLabel("\(day.day): \(formatActivityMetric(day.value, metric: metric))")
    }

    private func color(_ intensity: Int) -> Color {
        switch intensity {
        case 1: Theme.Colors.mutedForeground.opacity(0.2)
        case 2: Theme.Colors.mutedForeground.opacity(0.35)
        case 3: Theme.Colors.mutedForeground.opacity(0.55)
        case 4: Theme.Colors.foreground.opacity(0.8)
        default: Theme.Colors.selection.opacity(0.4)
        }
    }

    private var footer: String {
        if let selected {
            return
                "\(selected.date.formatted(.dateTime.month(.abbreviated).day())): \(formatActivityMetric(selected.value, metric: metric))"
        }
        let totals = activityContributionTotals(points, metric: metric)
        let value = formatActivityMetric(totals.visibleTotal, metric: metric)
        return metric == .activity ? "\(value) · \(totals.currentStreak) day streak" : value
    }

    private var description: LocalizedStringResource {
        switch metric {
        case .activity: "Agent starts and pull requests completed through Yiru."
        case .tokens: "Provider-reported token usage attributed to Yiru worktrees."
        case .value: "Standard global API-equivalent value calculated per request."
        }
    }
}
